File: Smart Sweepers/src/minesweeper/Minesweeper.java
package minesweeper;

import java.util.ArrayList;
import java.util.List;

import neuronalNet.NeuralNet;
import utils.C2DMatrix;
import utils.rand;
import utils.SPoint;
import utils.SVector2D;
import configuration.Parameters;

public class Minesweeper {

	private NeuralNet itsBrain;
	private SVector2D position;
	private SVector2D lookAt;

	private double rotation;
	private double speed;
	private double lTrack;
	private double rTrack;
	private double fitness;
	private double scale;

	int closestMine;

	public Minesweeper() {
		lTrack = 0.16;
		rTrack = 0.16;
		scale = Parameters.iSweeperScale;
		closestMine = 0;
		itsBrain = new NeuralNet();

		lookAt = new SVector2D(0, 0);

		reset();
	}

	public void reset() {
		fitness = 0;
		rotation = rand.randomFloat() * Math.PI * 2;
		position = new SVector2D(rand.randomFloat() * Parameters.WindowWidth, rand.randomFloat() * Parameters.WindowHeight);
	}

	public void worldTransform(List<SPoint> sweeper) {
		C2DMatrix matTransform = new C2DMatrix();

		matTransform.scale(scale, scale);

		matTransform.rotate(rotation);

		matTransform.translate(position.getX(), position.getY());

		matTransform.transformSPoints(sweeper);
	}

	public boolean update(List<SVector2D> mines) {
		List<Double> inputs = new ArrayList<>();

		SVector2D closestMine = getClosestMine(mines);

		closestMine = closestMine.normalize();

		inputs.add(closestMine.getX());
		inputs.add(closestMine.getY());

		inputs.add(lookAt.getX());
		inputs.add(lookAt.getY());

		List<Double> output = itsBrain.update(inputs);

		if (output.size() < Parameters.iNumOutputs) {
			return false;
		}

		lTrack = output.get(0);
		rTrack = output.get(1);

		double rotForce = lTrack - rTrack;

		rotForce = clamp(rotForce, -Parameters.dMaxTurnRate, Parameters.dMaxTurnRate);

		rotation += rotForce;

		speed = lTrack + rTrack;

		lookAt.setX(-Math.sin(rotation));
		lookAt.setY(Math.cos(rotation));

		position.plusAssign(lookAt.times(speed));

		if (position.getX() > Parameters.WindowWidth)
			position.setX(0);
		if (position.getX() < 0)
			position.setX(Parameters.WindowWidth);
		if (position.getY() > Parameters.WindowHeight)
			position.setY(0);
		if (position.getY() < 0)
			position.setY(Parameters.WindowHeight);

		return true;
	}

	private SVector2D getClosestMine(List<SVector2D> mines) {
		double closestSoFar = Double.MAX_VALUE;
		SVector2D closestObect = new SVector2D(0, 0);

		for (int i = 0; i < mines.size(); i++) {
			SVector2D currentMine = mines.get(i);
			double lenToOjbect = currentMine.minus(position).length();

			if (lenToOjbect < closestSoFar) {
				closestSoFar = lenToOjbect;
				closestObect = position.minus(currentMine);
				closestMine = i;
			}
		}
		return closestObect;
	}

	public int checkForMine(List<SVector2D> mines, double size) {
		SVector2D distanceToObject = position.minus(mines.get(closestMine));
		if (distanceToObject.length() < (size + 5)) {
			return closestMine;
		}

		return -1;
	}

	public SVector2D getPosition() {
		return position;
	}

	public void incrementFitness() {
		++fitness;
	}

	public double getFitness() {
		return fitness;
	}

	public void putWeights(List<Double> w) {
		itsBrain.putWeights(w);
	}

	public int getNumberOfWeights() {
		return itsBrain.getNumberOfWeights();
	}

	private double clamp(double arg, double min, double max) {
		if (arg < min) {
			return min;
		}
		if (arg > max) {
			return max;
		}
		return arg;
	}

}

File: Smart Sweepers/src/minesweeper/Minesweeper.kt
package minesweeper

import configuration.*
import neuronalNet.*
import utils.*
import utils.rand.randomFloat
import kotlin.math.*

class Minesweeper {
    private val itsBrain = NeuralNet()
    private lateinit var position: SVector2D
    private val lookAt = SVector2D(0.0, 0.0)

    private var rotation = 0.0
    private var speed = 0.0
    private var lTrack = 0.16
    private var rTrack = 0.16
    var fitness: Double = 0.0
        private set
    private val scale = Parameters.iSweeperScale.toDouble()

    private var closestMine: Int = 0

    init {
        reset()
    }

    fun reset() {
        fitness = 0.0
        rotation = rand.randomRadian()
        position = SVector2D(randomFloat() * Parameters.WindowWidth, randomFloat() * Parameters.WindowHeight)
    }

    fun worldTransform(sweeper: List<SPoint>) =
        with(C2DMatrix()) {
            scale(scale, scale)
            rotate(rotation)
            translate(position.x, position.y)
            transformSPoints(sweeper)
        }

    fun update(mines: List<SVector2D>): Boolean {
        val (between, found) = position.vectorToClosestOf(mines)
        closestMine = found
        val towardClosestMine = between.normalize()

        val output = itsBrain.update(listOf(towardClosestMine.x, towardClosestMine.y, lookAt.x, lookAt.y))

        lTrack = output[0]; rTrack = output[1]

        rotation += (lTrack - rTrack).boundedBy(-Parameters.dMaxTurnRate, Parameters.dMaxTurnRate)

        speed = lTrack + rTrack

        lookAt.x = -sin(rotation); lookAt.y = cos(rotation)

        position += lookAt * speed

        with(position) {
            x = x.wrappedTo(0.0, Parameters.WindowWidth.toDouble())
            y = y.wrappedTo(0.0, Parameters.WindowHeight.toDouble())
        }

        return true
    }

    fun checkForMine(mines: List<SVector2D>, size: Double) =
        if (position.closeEnoughForJazz(mines[closestMine], size)) closestMine
        else -1

    fun incrementFitness() {
        ++fitness
    }

    fun putWeights(w: List<Double>) =
        itsBrain.putWeights(w)

    val numberOfWeights: Int
        get() = itsBrain.numberOfWeights

    private companion object {
        /** if outside of bounds, returns nearest */
        fun <T : Comparable<T>> T.boundedBy(lower: T, upper: T) =
            if (this < lower) lower
            else if (this > upper) upper
            else this

        /** if outside of bounds, returns furthest */
        fun <T : Comparable<T>> T.wrappedTo(lower: T, upper: T) =
            if (this < lower) upper
            else if (this > upper) lower
            else this

        /** determines the nearest place in list, returns a vector *to* it, and its list-index */
        fun SVector2D.vectorToClosestOf(places: List<SVector2D>) =
            places
                .mapIndexed { i, place ->
                    val vectorBetween = this - place
                    val distanceBetween = vectorBetween.length()
                    Triple(distanceBetween, vectorBetween, i)
                }
                .minBy { it.first }
                .let { it.second to it.third }

        /** So if you are "close enough for Jazz" are you thus "On the Jazz"? */
        fun SVector2D.closeEnoughForJazz(place: SVector2D, size: Double) =
            (this - place).length() < size + 5
    }
}
